tutorials/iot/graphical-snaps/graphical-snaps.md

```diff
@@ -0,0 +1,608 @@
+```bash
+sudo usermod -a -G lxd $USER
+```
+
+Then sign out and back in (or `newgrp lxd` in the shell you'll be using)
+
+## Introducing glmark2-wayland
+duration: 5:00
+
+Let’s begin with a trivial example: `glmark2-wayland` - it is a test application that uses OpenGL and Wayland. This is a useful snap for verifying that the graphics stack of your hardware is correctly set up.
+
+### Install mir-demos
+When creating a graphical snap it is useful to be able to experiment on the desktop. For this we need to install the latest Mir, which requires a PPA:
+
+```bash
+sudo add-apt-repository ppa:mir-team/release
+sudo apt update
+sudo apt install mir-demos
+```
+
+### Install glmark2-wayland
+
+Next install glmark2-wayland:
+
+```bash
+sudo apt install glmark2-wayland
+```
+### Verify glmark2-wayland runs on Mir
+```bash
+miral-app -kiosk -launcher 'glmark2-wayland'
+```
+
+Inside the Mir-on-X window, you should see various graphical animations, and statistics printed to your console. All should be fine before proceeding.
+
+## First Pass Snapping: Test on Desktop
+duration: 5:00
+
+For our first pass we will snap glmark2-wayland and run it in DevMode (i.e. unconfined) on our Ubuntu desktop.
+
+This guide assumes you are familiar with creating snaps. If not, please read [here](https://docs.snapcraft.io/build-snaps/) first. Create the snap directory by forking https://github.com/snapcrafters/fork-and-rename-me
+
+```bash
+git clone https://github.com/snapcrafters/fork-and-rename-me.git glmark2-wayland
+```
+
+Inside the glmark2-wayland directory edit the “snapcraft.yaml” file, and let’s try the following (SPOILER, won’t work immediately, read on to troubleshoot):
+
+```yaml
+name: glmark2-wayland
+version: 0.1
+summary: GLMark2 on Wayland
+description: |
+  GLMark2 on Wayland
+confinement: strict
+grade: devel
+
+apps:
+  glmark2-wayland:
+    command: "usr/bin/glmark2-wayland"
+    plugs:
+      - opengl
+      - wayland
+
+parts:
+  glmark2-wayland:
+    plugin: nil
+    stage-packages:
+      - glmark2-wayland
+```
+
+Create the snap by returning to the “glmark2-wayland” directory and running
+```bash
+snapcraft cleanbuild
+```
+You should be left with a “glmark2-wayland_0.1_amd64.snap” file.
+
+Let’s test it!
+```bash
+miral-kiosk&
+sudo snap install --dangerous ./glmark2-wayland_0.1_amd64.snap --devmode
+snap run glmark2-wayland
+```
+(“dangerous” needed as local snap file being installed, and “devmode” required as we need to debug it)
+
+Oh no! It fails with these errors:
+```bash
+Error: Failed to open models directory '/usr/share/glmark2/models'
+Error: Failed to open models directory '/usr/share/glmark2/textures'
+Error: main: Could not initialize canvas
+```
+We need to solve these.
+
+## Common Problem 1: Files are not where they’re expected to be!
+duration: 5:00
+
+One important thing to remember about snaps is that all files are located in a subdirectory $SNAP which maps to /snap/<snap_name>/<version>. To prove this, try the following:
```

Comment:
This needs HTML enities for the greater/lesser signs `&gt; &lt;` (and probably a bug against the tutorial machinery)